### ISSUE TITLE
Make proxy port available in Consul

### ISSUE THREAD

**Issue by sgrimm-sg** [eBay/fabio#243]
In cases where Fabio sits in front of an internal microservice for load-balancing or failover purposes, there's currently no good way for clients of the microservice to dynamically discover the port number they should connect to. Fabio registers its administrative / health-check port with Consul, not its proxy port(s).

Two possible approaches come to mind (not necessarily mutually exclusive):

1. Let the administrator configure Fabio's proxy port(s) using the Consul KV store. Then an application could query this config setting. Obviously this needs to take into account the possibility of multiple Fabio instances with different configurations sharing the same Consul service.
2. Have Fabio register the current listen ports in Consul at startup. A tag on the Fabio service would work, or a KV entry.

**Comment by sgrimm-sg** [comment]
Or I guess 3. Have a machine-friendly query endpoint on the administrative listen port that a client can use to get a list of the proxy ports. Not ideal (more round trips required to discover the port number) but it would work.

**Comment by magiconair** [comment]
The reason I didn't consider this to be an issue is that fabio shouldn't be used for routing in the backend IMHO since consul already provides the necessary tools to do this. Couldn't you register your internal microservice in consul and have your other services discover it through the consul DNS interface?

If that isn't sufficient - and I am curious about your setup - then fabio should register its listeners as services in consul to advertise the ports IMO. I wouldn't rely on configuration since this has to be managed and that opens a lot of other issues. fabio has all the information and it can publish it. The question for me remains whether it needs to.

**Comment by sgrimm-sg** [comment]
Our use case is a bit more specific than the one I described in the issue. Fabio sits in front of a service that's used both by other components of our system (which can indeed discover the underlying service addresses directly in Consul) and third-party clients. Our application runs in AWS and the third-party clients connect through an ELB frontend that sits in front of Fabio.

The shutdown sequence from issue #242 is where I'd use this feature. When failing over to a new instance of the service I want to make sure that third-party clients are connecting to the new instance rather than the old one. Fabio is the source of truth there: if Fabio actually routes a request to the new instance, then the new instance is reachable and it's safe for the old one to deregister itself.

The ELB configuration needs to know the Fabio listen port, so there's no way to make the entire setup fully dynamic, but the fewer explicit config options I have to set the better, and this seems like the sort of thing that _could_ be discovered at runtime.

**Comment by magiconair** [comment]
yes, fabio could advertise its listening ports in consul or some other registry. However, the part that is missing is automating the ELB configuration based on that. So if I understand you correctly you would want to automate the ELB configuration at runtime so that when a fabio starts up it registers its listeners in consul and the magic code reconfigures the ELB to route traffic to the fabio instances. Correct?

**Comment by sean-** [comment]
What I'm about to suggest is for creativity purposes only and not something I'm recommending, but: if Fabio did advertise its ports back to Consul, it would be possible to Consul-template listen on the k/v, render out a Terraform config file locally, and then run Terraform to mutate the ports in an elb/alb. Locking, state sync, a whole host of issues come into play, but... if someone wanted to Rube-Goldberg up a solution it's possible.

**Comment by stephane-martin** [comment]
Let the administrator configure Fabio's proxy port(s) using the Consul KV store.

This is what i do so far: 

```bash
consul kv put XXX/fabio/FABIO_PROXY_ADDR 10.1.2.3:9999
/usr/local/bin/envconsul -prefix=XXX/fabio -pristine /bin/bash -c '/usr/local/bin/fabio -cfg /etc/fabio.conf'
```

And for the nginx in front of fabio:

```bash
/usr/local/bin/envconsul -prefix=XXX/fabio -upcase -pristine -sanitize /usr/local/bin/bash -c '/usr/local/bin/envsubst < /etc/nginx/nginx.conf.tmpl > /etc/nginx/nginx.conf \$FABIO_PROXY_ADDR && /usr/local/sbin/nginx -c /etc/nginx/nginx.conf'
```

`nginx.conf.tmpl` has something like:

```
upstream fabio_backend {
    server ${FABIO_PROXY_ADDR};
}
```

No really painful, but indeed it would be easier to have fabio registering itself.
